middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+const nearAPI = require('near-api-js');
+const crypto = require('crypto');
+const nacl = require('tweetnacl');
+const { creatorKeyJson } = require('./near');
+const { sendSms } = require('../utils/sms');
+const { sendMail } = require('../utils/email');
+const models = require('../models');
+const Sequelize = require('sequelize');
+const Op = Sequelize.Op;
+const password = require('secure-random-password');
+const SECURITY_CODE_DIGITS = 6;
+const twoFactorMethods = ['2fa-email', '2fa-phone'];
+const viewMethods = [];
+const changeMethods = ['confirm'];
+
+const getContract = async (contractName, secretKey) => {
+    const keyStore = {
+        async getKey() {
+            return nearAPI.KeyPair.fromString(secretKey);
+        },
+    };
+    const near = await nearAPI.connect({
+        deps: { keyStore },
+        masterAccount: creatorKeyJson && creatorKeyJson.account_id,
+        nodeUrl: process.env.NODE_URL
+    });
+    const contractAccount = new nearAPI.Account(near.connection, contractName);
+    const contract = new nearAPI.Contract(contractAccount, contractName, {
+        viewMethods,
+        changeMethods,
+    });
+    return contract;
+};
+const confirmRequest = async (accountId, request_id) => {
+    const key = await getDetermKey(accountId);
+    const contract = await getContract(accountId, key.secretKey);
+    // always parseInt on requestId. requestId is string in db because it could come from url params etc...
+    const res = await contract.confirm({ request_id: parseInt(request_id) }).catch((e) => {
+        return { success: false, error: e };
+    });
+    return { success: !!res, res };
+};
+/********************************
+2FA Key
+********************************/
+// helper: generates a deterministic key based on the accountId
+const getDetermKey = async (accountId) => {
+    const hash = crypto.createHash('sha256').update(accountId + creatorKeyJson.private_key).digest();
+    const keyPair = nacl.sign.keyPair.fromSeed(hash);//nacl.sign.keyPair.fromSecretKey(hash)
+    return {
+        publicKey: `ed25519:${nearAPI.utils.serialize.base_encode(keyPair.publicKey)}`,
+        secretKey: nearAPI.utils.serialize.base_encode(keyPair.secretKey)
+    };
+};
+/********************************
+Sending Codes
+method.kind = ['2fa-email', '2fa-phone']
+********************************/
+const prettyRequestInfo = ({ request }) => `
+    Transaction Recipient: ${ request.receiver_id }
+    Actions:\n\t${ request.actions.map((r) => r.type + (r.amount ? ': ' + nearAPI.utils.format.formatNearAmount(r.amount, 4) : '')).join('\n\t') }
```

Comment:
We need to have user-friendly formatting for actions.

@kcole16 may you help with copy for different type of tx actions? Should we basically match what wallet / explorer do?

Reply:
TBD